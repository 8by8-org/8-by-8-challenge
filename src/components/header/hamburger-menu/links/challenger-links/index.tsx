export { ChallengerLinks } from './challenger-links';
