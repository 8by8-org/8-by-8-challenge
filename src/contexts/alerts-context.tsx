'use client';
import { createNamedContext } from '@/hooks/create-named-context';
import { Alert, useAlert } from '@/components/utils/alert';
import type { PropsWithChildren } from 'react';

interface AlertsContextType {
  showAlert(message: string, variant: 'error' | 'success'): void;
}

export const AlertsContext =
  createNamedContext<AlertsContextType>('AlertsContext');

export function AlertsContextProvider({ children }: PropsWithChildren) {
  const { alertRef, showAlert } = useAlert();

  return (
    <AlertsContext.Provider value={{ showAlert }}>
      <Alert ref={alertRef} />
      {children}
    </AlertsContext.Provider>
  );
}
