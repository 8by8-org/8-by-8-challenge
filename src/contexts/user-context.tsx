'use client';
import { useState, useLayoutEffect, type PropsWithChildren } from 'react';
import { createNamedContext } from '../hooks/functions/create-named-context';
import { useContextSafely } from '@/hooks/functions/use-context-safely';
import { ServicesContext } from './services-context';
import type { User } from '../model/types/user';
import type { Avatar } from '@/model/types/avatar.type';
import type { UserType } from '@/model/enums/user-type';

interface UserContextType {
  user: User | null;
  signUpWithEmail(
    email: string,
    name: string,
    avatar: Avatar,
    type: UserType,
  ): Promise<void>;
  signInWithEmail(email: string): Promise<void>;
  signOut(): void;
}

const UserContext = createNamedContext<UserContextType>('UserContext');

function UserContextProvider({ children }: PropsWithChildren) {
  const { userService } = useContextSafely(
    ServicesContext,
    'UserContextProvider',
  );
  const [user, setUser] = useState<User | null>(userService.user);

  useLayoutEffect(() => {
    const subscription = userService.subscribe((user: User | null) =>
      setUser(user),
    );
    return subscription.unsubscribe;
  });

  return (
    <UserContext.Provider
      value={{
        signUpWithEmail: userService.signUpWithEmail,
        signInWithEmail: userService.signInWithEmail,
        signOut: userService.signOut,
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
}

export { UserContext, UserContextProvider, type UserContextType };
