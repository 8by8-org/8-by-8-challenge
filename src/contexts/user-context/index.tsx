export { UserContext, type UserContextType } from './user-context';
export { LocalUserContextProvider } from './local-user-context-provider';
