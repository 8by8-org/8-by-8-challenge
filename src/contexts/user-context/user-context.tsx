'use client';
import { createNamedContext } from '../../hooks/create-named-context';
import { VoterRegistrationForm } from '@/app/register/voter-registration-form';
import type { User } from '../../model/types/user';
import type { Avatar } from '@/model/types/avatar';
import type { ValueOf } from 'fully-formed';
import type { InvitedBy } from '@/model/types/invited-by';

interface SignUpWithEmailParams {
  email: string;
  name: string;
  avatar: Avatar;
  captchaToken: string;
}

interface SendOTPToEmailParams {
  email: string;
  captchaToken: string;
}

interface SignInWithOTPParams {
  otp: string;
}

interface UserContextType {
  user: User | null;
  emailForSignIn: string;
  invitedBy: InvitedBy | null;
  signUpWithEmail(params: SignUpWithEmailParams): Promise<void>;
  sendOTPToEmail(params: SendOTPToEmailParams): Promise<void>;
  resendOTP(): Promise<void>;
  signInWithOTP(params: SignInWithOTPParams): Promise<void>;
  signOut(): Promise<void>;
  restartChallenge(): Promise<void>;
  registerToVote(
    formData: ValueOf<InstanceType<typeof VoterRegistrationForm>>,
  ): Promise<void>;
  gotElectionReminders(): Promise<void>;
  shareChallenge(): Promise<void>; 
}

const UserContext = createNamedContext<UserContextType>('UserContext');

export {
  UserContext,
  type UserContextType,
  type SignUpWithEmailParams,
  type SendOTPToEmailParams,
  type SignInWithOTPParams,
};
