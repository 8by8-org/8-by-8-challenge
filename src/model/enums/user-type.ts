export enum UserType {
  Challenger = 'challenger',
  Player = 'player',
  Hybrid = 'hybrid',
}
