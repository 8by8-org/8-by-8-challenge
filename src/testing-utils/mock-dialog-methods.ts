/**
 * Mocks HTMLDialogElement methods as they are not supported by our test framework at this time.
 */
export function mockDialogMethods() {
  HTMLDialogElement.prototype.showModal = () => {};
  HTMLDialogElement.prototype.close = () => {};
}
