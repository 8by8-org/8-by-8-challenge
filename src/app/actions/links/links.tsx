'use client';
import React from 'react';
import Link from 'next/link';
import { useContextSafely } from '@/hooks/use-context-safely';
import { UserContext } from '@/contexts/user-context';
import { hasCompletedMultipleActions } from '../utils/has-completed-multiple-actions';
import { hasCompletedAllActions } from '../utils/has-completed-all-actions';
import { lastContributedToCurrentInviter } from '../utils/last-contributed-to-current-inviter';
import { UserType } from '@/model/enums/user-type';
import styles from './styles.module.scss';

export function Links() {
  const { user, invitedBy } = useContextSafely(UserContext, 'Links');

  if (hasCompletedAllActions(user)) {
    return null;
  }

  const links: React.JSX.Element[] = [];

  if (
    invitedBy &&
    lastContributedToCurrentInviter(user?.contributedTo, invitedBy)
  ) {
    if (hasCompletedMultipleActions(user)) {
      links.push(<p>Thanks for your actions!</p>);
      links.push(
        <Link href="/share" className="link--teal">
          Share about your actions
        </Link>,
      );
    } else if (
      user?.completedActions.electionReminders ||
      user?.completedActions.registerToVote
    ) {
      if (user?.completedActions.electionReminders) {
        links.push(<p>Thanks for getting election reminders!</p>);
      }

      if (user?.completedActions.registerToVote) {
        links.push(<p>Thanks for registering to vote!</p>);
      }

      links.push(
        <Link href="/share" className="link--teal">
          Share about your action
        </Link>,
      );
    } else if (user?.type === UserType.Hybrid) {
      links.push(<p>Thanks for taking the challenge!</p>);
    }
  } else if (user?.type === UserType.Hybrid) {
    links.push(<p>Here for something else?</p>);
  }

  if (user?.type === UserType.Hybrid) {
    links.push(
      <Link href="/progress" className="link--teal">
        See your progress
      </Link>,
    );
  }

  return <div className={styles.container}>{...links}</div>;
}
