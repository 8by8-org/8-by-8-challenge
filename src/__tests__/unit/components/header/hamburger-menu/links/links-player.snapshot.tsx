import { render, cleanup } from '@testing-library/react';
import { Builder } from 'builder-pattern';
import { UserContext, UserContextType } from '@/contexts/user-context';
import { AlertsContextProvider } from '@/contexts/alerts-context';
import { Links } from '@/components/header/hamburger-menu/links/links';
import { UserType } from '@/model/enums/user-type';
import {
  HeaderContext,
  type HeaderContextType,
} from '@/components/header/header-context';
import { mockDialogMethods } from '@/utils/test/mock-dialog-methods';
import type { User } from '@/model/types/user';

describe('Links--Player', () => {
  mockDialogMethods();

  afterEach(cleanup);

  it('renders player links unchanged', () => {
    const user = Builder<User>().type(UserType.Player).build();
    const headerCtxValue = Builder<HeaderContextType>().build();

    const { container } = render(
      <AlertsContextProvider>
        <UserContext.Provider value={{ user } as UserContextType}>
          <HeaderContext.Provider value={headerCtxValue}>
            <Links />
          </HeaderContext.Provider>
        </UserContext.Provider>
      </AlertsContextProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
