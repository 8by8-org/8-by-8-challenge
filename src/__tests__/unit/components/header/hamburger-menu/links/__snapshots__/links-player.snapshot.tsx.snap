// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Links--Player renders player links unchanged 1`] = `
<div>
  <div
    class="alert_container"
  >
    <div
      class="hidden"
      role="alert"
    />
  </div>
  <li>
    <button
      class="gradient sm take_the_challenge_btn"
    >
      <span>
        Take The Challenge
      </span>
    </button>
  </li>
  <li>
    <a
      class="link_lg_top"
      href="/actions"
    >
      Take Action
    </a>
  </li>
  <li>
    <a
      class="link_lg"
      href="/why8by8"
    >
      Why 8by8
    </a>
  </li>
  <li>
    <a
      class="link_sm_top"
      href="/tos"
    >
      Terms of Service
    </a>
  </li>
  <li>
    <a
      class="link_sm"
      href="/privacy"
    >
      Privacy Policy
    </a>
  </li>
  <li>
    <button
      class="signout_btn"
    >
      Sign out
    </button>
  </li>
</div>
`;
